enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/app_permissions.clj

```diff
@@ -0,0 +1,102 @@
+  (log/fatal "checking advanced permsissions")
+  (when-not (premium-features/has-feature? :advanced-permissions)
+    (throw (ex-info (i18n/tru "The granular app permission functionality is only enabled if you have a premium token with the advanced-permissions feature.")
+                    {:status-code 402}))))
+
+(defn- replace-collection-ids
+  "Convert the collection group permission graph `g` into an app group
+  permission graph or vice versa by removing the root collection and replacing
+  the collection IDs by app IDs according to `mapping`.
+
+  Note that the app permission graph never contains the root collection. So it's
+  removed when converting a collection graph to an app graph and plays no role
+  when converting an app graph to a collection graph."
+  [g mapping]
+  (update-vals g (fn [group-permissions]
+                   (-> group-permissions
+                       (dissoc :root)
+                       (update-keys mapping)))))
+
+(defn- merge-graphs
+  [base override]
+  (merge-with merge base override))
+
+(s/defn graph :- PermissionsGraph
+  "Returns the app permission graph. Throws an exception if the
+  advanced-permissions feature is not enabled."
+  []
+  (check-advanced-permissions)
+  (db/transaction
+   (let [collection-id->app-id (set/map-invert (db/select-id->field :collection_id 'App))]
+     (-> collection-id->app-id
+         keys
+         graph/collection-permission-graph
+         (update :groups replace-collection-ids collection-id->app-id)))))
+
+(defn update-collection-graph!
+  "Update the Collections permissions graph for Collections of `collection-namespace` (default `nil`, the 'default'
+  namespace). This works just like [[metabase.models.permission/update-data-perms-graph!]], but for Collections;
+  refer to that function's extensive documentation to get a sense for how this works."
+  [old-graph new-graph]
+  (let [old-perms          (:groups old-graph)
+        new-perms          (:groups new-graph)
+        ;; filter out any groups not in the old graph
+        new-perms          (select-keys new-perms (keys old-perms))
+        ;; filter out any collections not in the old graph
+        new-perms          (into {} (for [[group-id collection-id->perms] new-perms]
+                                      [group-id (select-keys collection-id->perms (keys (get old-perms group-id)))]))
+        [diff-old changes] (data/diff old-perms new-perms)]
+    (perms/log-permissions-changes diff-old changes)
+    (perms/check-revision-numbers old-graph new-graph)
+    (when (seq changes)
+      (db/transaction
+        (doseq [[group-id new-group-perms] changes
+                [collection-id new-perms] new-group-perms]
+          (graph/update-collection-permissions! nil group-id collection-id new-perms))
+       (perms/save-perms-revision! CollectionPermissionGraphRevision (:revision old-graph)
+                                   (assoc old-graph :namespace nil) changes)))))
```

Comment:
This is a simplified copy of the code in `collection.graph`. In the followup introducing the `app` collection namespace I'll make an attempt to reduce duplication.